README.md
# @himenon/git-control-js

Wrapper for using git commands in Node.js.

## Install

```
yarn add @himenon/git-control-js
```

## Usage

```ts
import * as GitControl from "@himenon/git-control-js";

const main = async () => {
  const git = GitControl.Wrap.create(process.cwd());
  await git.clone({
    owner: "Himenon",
    repo: "git-control-js",
    branch: "main",
    baseUrl: "https://github.com",
    baseSsh: "git@github.com",
    protocol: "https",
    outputDir: process.cwd(),
    authToken: "", // https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/#http-based-git-access-by-an-installation
  });

  const latestCommitDate = await git.getLatestCommitDate()
  console.log(latestCommitDate);
}

main().catch(error => {
  console.error(error);
  process.exit(1);
})
```

## Development

| scripts                   | description                                 |
| :------------------------ | :------------------------------------------ |
| `build`                   | typescript build and create proxy directory |
| `clean`                   | clean up                                    |
| `format:code`             | prettier                                    |
| `format:yarn:lock`        | yarn.lock deduplicate                       |
| `lerna:version:up`        | lerna version up                            |
| `test`                    | execute test:depcruise, test:jest           |
| `test:depcruise`          | dependency-cruiser's test                   |
| `test:jest`               | jest test                                   |
| `ts`                      | execute ts-node                             |
| `release:github:registry` | publish github registry                     |
| `release:npm:registry`    | publish npm registry                        |

## Features

- [Proxy Directory](https://himenon.github.io/docs/javascript/proxy-directory-design-pattern/)

## Release

- Automatic version updates are performed when merged into the `main` branch.

## LICENCE

[@himenon-git-control-js](https://github.com/Himenon/git-control-js)・MIT


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="0.6.4"></a>
## 0.6.4 (2020-12-27)


### Chores

* **deps:** bump node-notifier from 8.0.0 to 8.0.1 ([#19](https://github.com/Himenon/git-control-js/issues/19)) ([0eedbd2](https://github.com/Himenon/git-control-js/commit/0eedbd2))





<a name="0.6.3"></a>
## 0.6.3 (2020-12-15)


### Chores

* **deps:** bump ini from 1.3.5 to 1.3.8 ([#17](https://github.com/Himenon/git-control-js/issues/17)) ([ce30ff0](https://github.com/Himenon/git-control-js/commit/ce30ff0))





<a name="0.6.2"></a>
## 0.6.2 (2020-12-15)


### Documentation

* update description ([#18](https://github.com/Himenon/git-control-js/issues/18)) ([bfabc8e](https://github.com/Himenon/git-control-js/commit/bfabc8e))





<a name="0.6.1"></a>
## 0.6.1 (2020-11-20)


### Chores

* fetchを追加 ([#16](https://github.com/Himenon/git-control-js/issues/16)) ([47e5f82](https://github.com/Himenon/git-control-js/commit/47e5f82))





<a name="0.6.0"></a>
# 0.6.0 (2020-11-20)


### Features

* ioの削除とwrapperの作成 ([#15](https://github.com/Himenon/git-control-js/issues/15)) ([5740378](https://github.com/Himenon/git-control-js/commit/5740378))


### BREAKING CHANGES

* interfaceを大きく変更





<a name="0.5.2"></a>
## 0.5.2 (2020-11-18)


### Bug Fixes

* set remote origin ([#14](https://github.com/Himenon/git-control-js/issues/14)) ([0f9f490](https://github.com/Himenon/git-control-js/commit/0f9f490))





<a name="0.5.1"></a>
## 0.5.1 (2020-11-17)


### Bug Fixes

* arguments ([#13](https://github.com/Himenon/git-control-js/issues/13)) ([5851f1d](https://github.com/Himenon/git-control-js/commit/5851f1d))





<a name="0.5.0"></a>
# 0.5.0 (2020-11-17)


### Features

* add auth origin ([#12](https://github.com/Himenon/git-control-js/issues/12)) ([c1f07b5](https://github.com/Himenon/git-control-js/commit/c1f07b5))





<a name="0.4.1"></a>
## 0.4.1 (2020-11-17)


### Bug Fixes

* interface ([#11](https://github.com/Himenon/git-control-js/issues/11)) ([915f632](https://github.com/Himenon/git-control-js/commit/915f632))





<a name="0.4.0"></a>
# 0.4.0 (2020-11-17)


### Features

* update interface ([#10](https://github.com/Himenon/git-control-js/issues/10)) ([dcda3f1](https://github.com/Himenon/git-control-js/commit/dcda3f1))





<a name="0.3.0"></a>
# 0.3.0 (2020-11-16)


### Features

* add set config ([#9](https://github.com/Himenon/git-control-js/issues/9)) ([480c3ca](https://github.com/Himenon/git-control-js/commit/480c3ca))





<a name="0.2.0"></a>
# 0.2.0 (2020-11-16)


### Features

* rename repoConfig -> config ([#8](https://github.com/Himenon/git-control-js/issues/8)) ([0e7f548](https://github.com/Himenon/git-control-js/commit/0e7f548))





<a name="0.1.1"></a>
## 0.1.1 (2020-11-15)


### Bug Fixes

* interface ([#6](https://github.com/Himenon/git-control-js/issues/6)) ([c3388ba](https://github.com/Himenon/git-control-js/commit/c3388ba))





<a name="0.1.0"></a>
# 0.1.0 (2020-11-15)


### Features

* rename ([#5](https://github.com/Himenon/git-control-js/issues/5)) ([0412912](https://github.com/Himenon/git-control-js/commit/0412912))


src/command.ts
import { EOL } from "os";
import * as Shell from "./shell";
import * as path from "path";
import { generateHttpBaseAccessUrl } from "./utils";
import { existDirectory } from "./filesystem";

export { Shell };

export interface CloneOption {
  owner: string;
  repo: string;
  branch: string;
  baseUrl: "https://github.com" | string;
  baseSsh: "git@github.com" | string;
  protocol: "ssh" | "https";
  outputDir: string;
  authToken?: string;
}

export interface SetConfigOption {
  key: string;
  value: string;
  type?: "local" | "global" | "system";
}

export interface CliOption {
  cwd?: string;
}

export interface GitRevListOption {
  remotes?: string;
  branches?: string;
}

export interface UpdateRemoteOption {
  owner: string;
  repo: string;
  remote: string;
  authToken: string;
}

export interface PushOption {
  remote: string;
  branch: string;
}

export interface CommitOption {
  shortMessage: string;
}

export interface AddOption {
  all?: boolean;
  file?: string;
}

export interface FetchOption {
  target: string;
}

export interface Type<T> {
  /**
   * git config --${type} ${key} ${value}
   */
  setConfig: (option: SetConfigOption) => T;
  /**
   *
   */
  updateRemote: (option: UpdateRemoteOption) => T;
  /**
   * 現在のローカルブランチ名を取得する
   * @example "master"
   */
  getBranch: (option: CliOption) => T;
  /**
   * 最新のコミットの日付取得する
   * @example "Mon Oct 28 20:18:16 2019 +0900"
   */
  getLatestCommitDate: (option: CliOption) => T;
  /**
   * Head CommitのShaを取得する
   */
  getHeadCommitSha: (option: CliOption) => T;
  /**
   * `git clone`
   *
   * @param owner オーナー
   * @param repo リポジトリ名
   * @param outputDir 出力先
   * @param branch ブランチ名
   * @param baseUrl GitHub EnterPriseなどのBaseUrl
   */
  clone: (option: CloneOption) => T;
  /**
   * git rev-list HEAD
   */
  getRevListHead: (params: GitRevListOption) => T;
  /**
   * `git status`
   */
  getStatus: (option?: CliOption) => T;
  /**
   * `git push`
   * @param remote
   * @param branch
   */
  push: (option: PushOption) => T;
  /**
   * `git commit -m ${message} --no-verify`
   * @param message
   */
  commit: (option: CommitOption) => T;
  /**
   * `git add -A`
   */
  add: (option: AddOption) => T;
  /**
   * git fetch [target]
   */
  fetch: (option: FetchOption) => T;
}

/**
 * gitコマンドを生成する
 * @param pwd コマンド実行ディレクトリ
 */
export const create = (workingDir: string): Type<Shell.Type> => {
  /**
   * @param args `git`コマンドの引数
   */
  const git = (args: string, cwd = workingDir) => {
    process.stdout.write("Command: " + "git " + args + EOL);
    return Shell.exec("git " + args, cwd);
  };
  return {
    setConfig: ({ key, value, type = "local" }) => {
      return git(`config --${type} ${key} ${value}`);
    },
    updateRemote: ({ owner, repo, remote, authToken }) => {
      const url = generateHttpBaseAccessUrl({ authToken, owner, repo });
      try {
        return git(`remote set-url ${remote} ${url}`);
      } catch (error) {
        git(`remote rm ${remote}`);
        return git(`remote add ${remote} ${url}`);
      }
    },
    getRevListHead: (option?: GitRevListOption): Shell.Type => {
      const options: string[] = [];
      const command = "rev-list HEAD";
      if (!option) {
        return git(command);
      }
      if (option.branches) {
        options.push(`--branches="${option.branches}"`);
      }
      if (option.remotes) {
        options.push(`--remotes="${option.remotes}"`);
      }
      return git([command, ...options].join(" "));
    },
    getBranch: ({ cwd }): Shell.Type => git("symbolic-ref --short HEAD", cwd),
    getLatestCommitDate: ({ cwd }): Shell.Type => git(`log --pretty=format:"%ad" -1`, cwd),
    getHeadCommitSha: ({ cwd }): Shell.Type => git("rev-parse HEAD", cwd),
    clone: ({ owner, repo, branch, baseUrl, baseSsh, protocol, outputDir, authToken }) => {
      if (existDirectory(outputDir)) {
        throw new Error(`Already exist directory: "${outputDir}`);
      }
      if (protocol === "ssh") {
        return git(`clone -b ${branch} ${baseSsh}:${owner}/${repo} ${outputDir}`, path.dirname(outputDir));
      }
      if (authToken) {
        const url = generateHttpBaseAccessUrl({ authToken, owner, repo });
        return git(`clone -b ${branch} ${url} ${outputDir}`, path.dirname(outputDir));
      }
      return git(`clone -b ${branch} ${baseUrl}/${owner}/${repo} ${outputDir}`, path.dirname(outputDir));
    },
    getStatus: (): Shell.Type => git("status"),
    push: ({ remote, branch }: PushOption): Shell.Type => git(`push ${remote} ${branch}`),
    commit: ({ shortMessage }: CommitOption): Shell.Type => git(`commit -m '${shortMessage}' --no-verify`),
    add: ({ all, file }: AddOption) => {
      if (all) {
        return git("add -A");
      }
      if (file) {
        return git(`add ${file}`);
      }
      return git("add -A");
    },
    fetch: ({ target }) => git(`fetch ${target}`),
  };
};
